src/util/custom-table.lzz

```diff
@@ -0,0 +1,329 @@
+		delete[] args;
+
+		if (maybe_return_value.IsEmpty()) {
+			self->PropagateJSError();
+			return SQLITE_ERROR;
+		}
+
+		v8::Local<v8::Array> returnValue = v8::Local<v8::Array>::Cast(maybe_return_value.ToLocalChecked());
+		v8::Local<v8::String> sqlString = v8::Local<v8::String>::Cast(returnValue->Get(ctx, 0).ToLocalChecked());
+		v8::Local<v8::Function> generator = v8::Local<v8::Function>::Cast(returnValue->Get(ctx, 1).ToLocalChecked());
+		v8::Local<v8::Array> parameterNames = v8::Local<v8::Array>::Cast(returnValue->Get(ctx, 2).ToLocalChecked());
+		int safe_ints = v8::Local<v8::Int32>::Cast(returnValue->Get(ctx, 3).ToLocalChecked())->Value();
+		bool direct_only = v8::Local<v8::Boolean>::Cast(returnValue->Get(ctx, 4).ToLocalChecked())->Value();
+
+		v8::String::Utf8Value sql(isolate, sqlString);
+		safe_ints = safe_ints < 2 ? safe_ints : static_cast<int>(self->db->GetState()->safe_ints);
+
+		std::vector<std::string> parameter_names;
+		int parameter_count = parameterNames->Length();
+		for (int i = 0; i < parameter_count; ++i) {
+			v8::Local<v8::String> parameterName = v8::Local<v8::String>::Cast(parameterNames->Get(ctx, i).ToLocalChecked());
+			v8::String::Utf8Value parameter_name(isolate, parameterName);
+			parameter_names.emplace_back(*parameter_name);
+		}
+
+		if (sqlite3_declare_vtab(db_handle, *sql) != SQLITE_OK) {
+			*errOutput = sqlite3_mprintf("failed to declare virtual table \"%s\"", argv[2]);
+			return SQLITE_ERROR;
+		}
+		if (direct_only && sqlite3_vtab_config(db_handle, SQLITE_VTAB_DIRECTONLY) != SQLITE_OK) {
+			*errOutput = sqlite3_mprintf("failed to configure virtual table \"%s\"", argv[2]);
+			return SQLITE_ERROR;
+		}
+
+		*output = reinterpret_cast<sqlite3_vtab*>(new VTab(self, parameter_count, safe_ints, generator, parameter_names));
+		return SQLITE_OK;
+	}
+
+	static int xDisconnect(sqlite3_vtab* vtab) {
+		delete reinterpret_cast<VTab*>(vtab);
+		return SQLITE_OK;
+	}
+
+	static int xOpen(sqlite3_vtab* vtab, sqlite3_vtab_cursor** output) {
+		*output = reinterpret_cast<sqlite3_vtab_cursor*>(new Cursor());
+		return SQLITE_OK;
+	}
+
+	static int xClose(sqlite3_vtab_cursor* cursor) {
+		delete reinterpret_cast<Cursor*>(cursor);
+		return SQLITE_OK;
+	}
+
+	static int xFilter(sqlite3_vtab_cursor* _cursor, int idxNum, const char* idxStr, int argc, sqlite3_value** argv) {
+		Cursor* cursor = reinterpret_cast<Cursor*>(_cursor);
+		VTab* vtab = reinterpret_cast<VTab*>(cursor->base.pVtab);
+		CustomTable* self = vtab->parent;
+		Addon* addon = self->addon;
+		v8::Isolate* isolate = self->isolate;
+		v8::HandleScope scope(isolate);
+		UseContext;
+
+		v8::Local<v8::Value> args_fast[4];
+		v8::Local<v8::Value>* args = NULL;
+		int parameter_count = vtab->parameter_count;
+		if (parameter_count != 0) {
+			args = parameter_count <= 4 ? args_fast : ALLOC_ARRAY<v8::Local<v8::Value>>(parameter_count);
```

Comment:
I think every single vtab test has 3 or less `rows` arguments and is always hitting that code path. Maybe add a test with more arguments.

Reply:
Good catch. Adding a test.